src/main.cpp

```diff
@@ -912,16 +912,121 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+                        }
+                    }
+                }
+
+                // Make sure the outputs of the transaction we're replacing
+                // have not been spent.
+                for (unsigned int j = 0; j < ptxConflicting->vout.size(); j++)
+                {
+                    if (pool.mapNextTx.count(COutPoint(hashConflicting, j)))
+                    {
+                        return state.DoS(0, error("AcceptToMemoryPool: outpoint %s:%d already spent; can't replace with %s",
+                                                  hashConflicting.ToString(), j,
+                                                  hash.ToString()),
+                                         REJECT_DUPLICATE, "bad-txns-inputs-spent");
+
+                    }
+                }
+
+                set<COutPoint> setConflictsPrevouts;
+                for (unsigned int j = 0; j < ptxConflicting->vin.size(); j++)
+                {
+                    setConflictsPrevouts.insert(ptxConflicting->vin[j].prevout);
+                }
+
+                for (unsigned int j = 0; j < tx.vin.size(); j++)
+                {
+                    // Check that any new inputs spend only confirmed coins. It
+                    // might not be economically rational to replace a
+                    // transaction that spent confirmed inputs with one that
+                    // spent the result of a long chain of inputs, as the total
+                    // fee/KB might go down. Forcing all new inputs to be
+                    // confirmed is a simple way of avoiding that problem.
+                    if (!setConflictsPrevouts.count(tx.vin[j].prevout))
+                    {
+                        // Rather than check the UTXO set - potentially
+                        // expensive - it's cheaper to just check that the new
+                        // input refers to a tx that is *not* in the mempool.
+                        if (pool.exists(tx.vin[j].prevout.hash))
+                            return state.DoS(0, error("AcceptToMemoryPool: replacement %s adds unconfirmed input, idx %d",
+                                                      hash.ToString(), j),
+                                             REJECT_DUPLICATE, "bad-txns-inputs-spent");
+                    }
+
+                    // Check that this transaction doesn't itself spend an output
+                    // of the conflicting transaction. If it did, when the conflict
+                    // was removed this transaction would become an orphan. As we
+                    // checked that the conflicts outputs were all unspent above,
+                    // we only have to check against the conflict, not any
+                    // children. Similarly because we'll only ever replace
+                    // transactions on a 1-1 basis, we only have to do this check
+                    // once.
+                    if (tx.vin[j].prevout.hash == hashConflicting)
```

Comment:
Ah, yeah, that's true now that it's been changed to require new inputs to be confirmed. Good catch.